Mathlib/NumberTheory/SmoothNumbers.lean

```diff
@@ -0,0 +1,150 @@
+
+lemma not_mem_primesBelow (n : ℕ) : n ∉ primesBelow n :=
+  fun hn ↦ (lt_of_mem_primesBelow hn).false
+
+/-- `smoothNumbers n` is the set of *`n`-smooth positive natural numbers*, i.e., the
+positive natural numbers all of whose prime factors are less than `n`. -/
+def smoothNumbers (n : ℕ) : Set ℕ := {m | m ≠ 0 ∧ ∀ p ∈ factors m, p < n}
+
+lemma mem_smoothNumbers {n m : ℕ} : m ∈ smoothNumbers n ↔ m ≠ 0 ∧ ∀ p ∈ factors m, p < n :=
+  Iff.rfl
+
+@[simp]
+lemma smoothNumbers_zero : smoothNumbers 0 = {1} := by
+  ext m
+  rw [Set.mem_singleton_iff, mem_smoothNumbers]
+  simp_rw [not_lt_zero]
+  rw [← List.eq_nil_iff_forall_not_mem, factors_eq_nil, and_or_left, not_and_self_iff, false_or,
+    ne_and_eq_iff_right zero_ne_one]
+
+/-- The product of the prime factors of `n` that are less than `N` is an `N`-smooth number. -/
+lemma prod_mem_smoothNumbers (n N : ℕ) : (n.factors.filter (· < N)).prod ∈ smoothNumbers N := by
+  have h₀ : (n.factors.filter (· < N)).prod ≠ 0 :=
+    List.prod_ne_zero fun h ↦ (pos_of_mem_factors (List.mem_of_mem_filter h)).false
+  refine ⟨h₀, fun p hp ↦ ?_⟩
+  obtain ⟨H₁, H₂⟩ := (mem_factors h₀).mp hp
+  simpa only [decide_eq_true_eq] using List.of_mem_filter <| mem_list_primes_of_dvd_prod H₁.prime
+    (fun _ hq ↦ (prime_of_mem_factors (List.mem_of_mem_filter hq)).prime) H₂
+
+/-- The sets of `N`-smooth and of `(N+1)`-smooth numbers are the same when `N` is not prime.
+See `Nat.equivProdNatSmoothNumbers` for when `N` is prime. -/
+lemma smoothNumbers_succ {N : ℕ} (hN : ¬ N.Prime) : N.succ.smoothNumbers = N.smoothNumbers := by
+  ext m
+  refine ⟨fun hm ↦ ⟨hm.1, fun p hp ↦ ?_⟩,
+         fun hm ↦ ⟨hm.1, fun p hp ↦ (hm.2 p hp).trans <| lt.base N⟩⟩
+  exact lt_of_le_of_ne (lt_succ.mp <| hm.2 p hp) fun h ↦ hN <| h ▸ prime_of_mem_factors hp
+
+@[simp] lemma smoothNumbers_one : smoothNumbers 1 = {1} := by simp [smoothNumbers_succ]
+
+@[gcongr] lemma smoothNumbers_mono {N M : ℕ} (hNM : N ≤ M) : N.smoothNumbers ⊆ M.smoothNumbers :=
+  fun _ hx ↦ ⟨hx.1, fun p hp => (hx.2 p hp).trans_le hNM⟩
+
+/-- The non-zero non-`N`-smooth numbers are `≥ N`. -/
+lemma smoothNumbers_compl (N : ℕ) : (N.smoothNumbers)ᶜ \ {0} ⊆ {n | N ≤ n} := by
+  intro n hn
+  simp only [Set.mem_compl_iff, mem_smoothNumbers, Set.mem_diff, ne_eq, not_and, not_forall,
+    not_lt, exists_prop, Set.mem_singleton_iff] at hn
+  obtain ⟨m, hm₁, hm₂⟩ := hn.1 hn.2
+  exact hm₂.trans <| le_of_mem_factors hm₁
+
+/-- If `p` is positive and `n` is `p`-smooth, then every product `p^e * n` is `(p+1)`-smooth. -/
+lemma pow_mul_mem_smoothNumbers {p n : ℕ} (hp : p ≠ 0) (e : ℕ) (hn : n ∈ smoothNumbers p) :
+    p ^ e * n ∈ smoothNumbers (succ p) := by
+  have hp' := pow_ne_zero e hp
+  refine ⟨mul_ne_zero hp' hn.1, fun q hq ↦ ?_⟩
+  rcases (mem_factors_mul hp' hn.1).mp hq with H | H
+  · rw [mem_factors hp'] at H
+    exact lt_succ.mpr <| le_of_dvd hp.bot_lt <| Prime.dvd_of_dvd_pow H.1 H.2
+  · exact (hn.2 q H).trans <| lt.base p
```

Comment:
i've never seen `lt.base` used; is there a better lemma for it?

Reply:
I guess this is what `exact?` gives you. Will replace by [lt_succ_self](https://leanprover-community.github.io/mathlib4_docs/Init/Data/Nat/Basic.html#Nat.lt_succ_self).